### PR TITLE
reduce operations on healthy machine

### PR DESCRIPTION
**Pull request by goushicui**
✨ (✨, feature additions) reduce operations on healthy machine

What this PR does / why we need it:
The controller will do the following two things to healthy machines, when any node updated

1.  Find out whether the ExternalRemediation Object exists, and delete it if it exists
2. Try patch the health check succeed condition on the machine
At this time, when the number of machines is very large, it is a waste of time.
According to logic, if Machine is always healthy, there is no need to do the above events, and it will also reduce the pressure on the API server.

### CODE REVIEW

File: controllers/machinehealthcheck_controller.go [hunk 1]
```diff
@@ -351,7 +355,8 @@ func (r *MachineHealthCheckReconciler) patchHealthyTargets(ctx context.Context,
 				}
 			}
 		}
-
+		// mark machine in healthy status after delete externalRemediation object
+		conditions.MarkTrue(t.Machine, clusterv1.MachineHealthCheckSuccededCondition)
```
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:359]: Moving the condition set here seems like a good improvements from a quick look, given that `healthCheckTargets` should have been read only from what I can gather from the original code comments.

cc @JoelSpeed

Comment (vincepri) [reply controllers/machinehealthcheck_controller.go:359]:
```suggestion
		// Mark machine in healthy status if it's healthy.
		conditions.MarkTrue(t.Machine, clusterv1.MachineHealthCheckSuccededCondition)
```

Not every Machine has a RemediationTemplate attached to it

File: controllers/machinehealthcheck_controller.go [hunk 2]
```diff
@@ -333,6 +333,10 @@ func (r *MachineHealthCheckReconciler) reconcile(ctx context.Context, logger log
 func (r *MachineHealthCheckReconciler) patchHealthyTargets(ctx context.Context, logger logr.Logger, healthy []healthCheckTarget, m *clusterv1.MachineHealthCheck) []error {
 	errList := []error{}
 	for _, t := range healthy {
+		// machine always in healthy status
+		if conditions.IsTrue(t.Machine, clusterv1.MachineHealthCheckSuccededCondition) {
+			continue
+		}
```
Comment (vincepri) [inline controllers/machinehealthcheck_controller.go:339]: 🤔 Would this check make the whole reconciliation non re-entrant? What happens if there is a remediation request but the machine was somehow marked as MachineHealthCheckSuccededCondition?

**Review by vincepri — COMMENTED** [review]
Hi @goushicui and thanks for the PR! I took a look briefly at the changes and the PR description, although could we have a bit more information why is this needed?

The PR title says we're trying to reduce operations on healthy machine, is the reduction significant in terms of performance, or was this done for a different reason? Usuallissuean having Usually having a more detailed companion issue to a PR can also help reviewers and maintainers.

File: controllers/machinehealthcheck_controller.go (hunk 1 above)
Comment (goushicui) [reply controllers/machinehealthcheck_controller.go:359]: Whether there is no RemediationTemplate or not, as long as the node is judged to be healthy, the healthy condition on the machine patch shall be provided

Comment (enxebre) [reply controllers/machinehealthcheck_controller.go:359]: Right, suggestions is to update comment to reflect that.

File: controllers/machinehealthcheck_targets.go [hunk 3]
```diff
@@ -302,7 +302,6 @@ func (r *MachineHealthCheckReconciler) healthCheckTargets(targets []healthCheckT
 		}
 
 		if t.Machine.DeletionTimestamp.IsZero() && t.Node != nil {
```
Comment (enxebre) [inline controllers/machinehealthcheck_targets.go:304]: `healthCheckTargets` marks MachineHealthCheckSuccededCondition as false but not as true now can we clarify that in `healthCheckTargets` comment elaborating why.

File: controllers/machinehealthcheck_controller.go (hunk 2 above)
Comment (enxebre) [reply controllers/machinehealthcheck_controller.go:339]: Right, the way this is in the PR if e.g [deleting the remediation request fails](https://github.com/kubernetes-sigs/cluster-api/pull/5729/files#diff-34cc13ef9d81a69fe97139ffce42c13fbf1ce1d330b00db41f9b29fcde9baaa4R353)
we set `MachineHealthCheckSuccededCondition` as true anyway and the next iteraction we don't retry to delete

**Comment by vincepri** [comment]
/close

Closing due to inactivity, please feel free to reopen when ready

**Pull request closed without merging**